llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -0,0 +1,712 @@
+  Or(Pred &&p, Preds &&...preds)
+      : Or<Preds...>(std::forward<Preds>(preds)...), P(std::forward<Pred>(p)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return P.match(Ctx, N) || Or<Preds...>::match(Ctx, N);
+  }
+};
+
+template <typename... Preds> And<Preds...> m_all_of(Preds &&...preds) {
+  return And<Preds...>(std::forward<Preds>(preds)...);
+}
+
+template <typename... Preds> Or<Preds...> m_any_of(Preds &&...preds) {
+  return Or<Preds...>(std::forward<Preds>(preds)...);
+}
+
+// === Generic node matching ===
+template <typename... OpndPreds> struct Node_match {
+  unsigned Opcode;
+  unsigned OpIdx;
+
+  Node_match(unsigned Opc, unsigned OpIdx) : Opcode(Opc), OpIdx(OpIdx) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (!N)
+      return false;
+
+    if (OpIdx == 0) {
+      // Check opcode
+      if (!sd_context_match(N, Ctx, m_Opc(Opcode)))
+        return false;
+    }
+
+    // Returns false if there are more operands than predicates;
+    return N->getNumOperands() == OpIdx;
+  }
+};
+
+template <typename OpndPred, typename... OpndPreds>
+struct Node_match<OpndPred, OpndPreds...> : Node_match<OpndPreds...> {
+  unsigned Opcode;
+  unsigned OpIdx;
+  OpndPred P;
+
+  Node_match(unsigned Opc, unsigned OpIdx, OpndPred &&p, OpndPreds &&...preds)
+      : Node_match<OpndPreds...>(Opc, OpIdx + 1,
+                                 std::forward<OpndPreds>(preds)...),
+        Opcode(Opc), OpIdx(OpIdx), P(std::forward<OpndPred>(p)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (!N)
+      return false;
+
+    if (OpIdx == 0) {
+      // Check opcode
+      if (!sd_context_match(N, Ctx, m_Opc(Opcode)))
+        return false;
+    }
+
+    if (OpIdx < N->getNumOperands())
+      return P.match(Ctx, N->getOperand(OpIdx)) &&
+             Node_match<OpndPreds...>::match(Ctx, N);
+
+    // This is the case where there are more predicates than operands.
+    return false;
+  }
+};
+
+template <typename... OpndPreds>
+Node_match<OpndPreds...> m_Node(unsigned Opcode, OpndPreds &&...preds) {
+  return Node_match<OpndPreds...>(Opcode, 0, std::forward<OpndPreds>(preds)...);
+}
+
+/// Provide number of operands that are not chain or glue, as well as the first
+/// index of such operand.
+struct EffectiveOperands {
+  unsigned Size = 0;
+  unsigned FirstIndex = 0;
+
+  explicit EffectiveOperands(SDValue N) {
+    const unsigned TotalNumOps = N->getNumOperands();
+    FirstIndex = TotalNumOps;
+    for (unsigned i = 0; i < TotalNumOps; ++i) {
+      // Count the number of non-chain and non-glue nodes (we ignore chain
+      // and glue by default) and retreive the operand index offset.
+      EVT VT = N->getOperand(i).getValueType();
+      if (VT != MVT::Glue && VT != MVT::Other) {
+        ++Size;
+        if (FirstIndex == TotalNumOps)
+          FirstIndex = i;
+      }
+    }
+  }
+};
```

Comment:
This is going to impose a significant cost on all matches, even though the "effective operand" handling is usually not relevant (for stuff like ADDs, etc). I think this should be opt-in, or possibly removed entirely (match helpers can add wildcards at the appropriate positions).

Reply:
I choose to keep it and allow users to opt in via `m_ChainedBinOp`, as I still want to match some binary operations like `STRICT_F*` with this.